src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -439,134 +378,251 @@ internal override void WriteAsPropertyNameCore(Utf8JsonWriter writer, T value, J
+                case TypeCode.Int32 or TypeCode.UInt32:
+                    uint uintValue = (uint)value;
+                    return Unsafe.As<uint, T>(ref uintValue);
+
+                case TypeCode.Int64 or TypeCode.UInt64:
+                    ulong ulongValue = value;
+                    return Unsafe.As<ulong, T>(ref ulongValue);
+
+                case TypeCode.Int16 or TypeCode.UInt16:
+                    ushort ushortValue = (ushort)value;
+                    return Unsafe.As<ushort, T>(ref ushortValue);
+
+                default:
+                    Debug.Assert(s_enumTypeCode is TypeCode.SByte or TypeCode.Byte);
+                    byte byteValue = (byte)value;
+                    return Unsafe.As<byte, T>(ref byteValue);
+            };
         }
 
-        private static string FormatJsonName(string value, JsonNamingPolicy? namingPolicy)
+        /// <summary>
+        /// Attempt to format the enum value as a comma-separated string of flag values, or returns false if not a valid flag combination.
+        /// </summary>
+        private bool TryFormatEnumAsString(ulong key, T value, JsonNamingPolicy? dictionaryKeyPolicy, [NotNullWhen(true)] out string? stringValue)
         {
-            if (namingPolicy is null)
-            {
-                return value;
-            }
+            Debug.Assert(!Enum.IsDefined(typeof(T), value) || dictionaryKeyPolicy != null, "Must either be used on undefined values or with a key policy.");
 
-            string converted;
-            if (!value.Contains(ValueSeparator))
+            if (s_isFlagsEnum)
             {
-                converted = namingPolicy.ConvertName(value);
-                if (converted == null)
+                using ValueStringBuilder sb = new(stackalloc char[JsonConstants.StackallocCharThreshold]);
+                ulong remainingBits = key;
+
+                foreach (EnumFieldInfo enumField in _enumFieldInfo)
                 {
-                    ThrowHelper.ThrowInvalidOperationException_NamingPolicyReturnNull(namingPolicy);
+                    ulong fieldKey = enumField.Key;
+                    if (fieldKey == 0 ? key == 0 : (remainingBits & fieldKey) == fieldKey)
+                    {
+                        string name = dictionaryKeyPolicy is not null
+                            ? ResolveAndValidateJsonName(enumField.Name, dictionaryKeyPolicy, enumField.IsNameFromAttribute)
+                            : enumField.JsonName;
+
+                        if (sb.Length > 0)
+                        {
+                            sb.Append(", ");
+                        }
+
+                        sb.Append(name);
+                        remainingBits &= ~fieldKey;
+
+                        if (fieldKey == 0)
+                        {
+                            // Do not process further fields if the value equals zero.
+                            Debug.Assert(key == 0);
+                            break;
+                        }
```

Comment:
That could exit early if your original value is zero and you're searching for a label that is zero. This could be added to the check of course, but given that we expect the array to be small it's a trade-off between fewer iteration vs. most checks per iteration. I'll see if I can pick a winner using microbenchmarks.

Reply:
> That could exit early if your original value is zero and you're searching for a label that is zero.

If your value is zero, it will only match with a label that is zero the same way the previous code did. I don't see how it would exit early.

Reply:
I've reworked the code to closer match what you are describing.